src/coordinator/storage/local/storage.go

```diff
@@ -65,14 +68,64 @@ func (s *localStorage) Fetch(ctx context.Context, query *storage.FetchQuery, opt
 		return nil, err
 	}
 
-	opts := storage.FetchOptionsToM3Options(options, query)
+	// NB(r): Since we don't use a single index we fan out to each
+	// cluster that can completely fulfill this range and then prefer the
+	// highest resolution (most fine grained) results.
+	// This needs to be optimized, however this is a start.
+	var (
+		opts       = storage.FetchOptionsToM3Options(options, query)
+		namespaces = s.clusters.ClusterNamespaces()
+		now        = time.Now()
+		fetches    = 0
+		result     multiFetchResult
+		wg         sync.WaitGroup
+	)
+	for _, namespace := range namespaces {
+		namespace := namespace // Capture var
+
+		clusterStart := now.Add(-1 * namespace.Attributes().Retention)
+
+		// Only include if cluster can completely fulfill the range
+		if clusterStart.After(query.Start) {
+			continue
+		}
+
+		fetches++
+
+		wg.Add(1)
+		go func() {
+			r, err := s.fetch(namespace, m3query, opts)
```

Comment:
do we want to exit on first error ? If yes, then probably use ExecuteParallel, if not, then probably add a comment since we shortcircuit on first error in most cases.

Reply:
I'd have to create temporary objects (i.e. allocation per fanout thing) which is costly, that's why I use a single static waitgroup and only if there are errors does it allocate a per fanout element.

Reply:
I see, so its cheaper to do everything in parallel instead of failing on first error and needing allocations ? I suppose I can change the other functions later then to stop using ExecuteParallel

Reply:
Basically cheapest is zero allocation if possible, though this is not always possible, everywhere that is a new allocation and then shared (i.e. escapes to heap) is a place that probably should be pooled if in the hot path.